File: Lab 1 source/CS5542_Lab1_PYTHON/movieRank.py
from pyspark import SparkContext
from pyspark import SparkConf

if __name__ == "__main__":

    # init spark program
    conf = SparkConf().setMaster("local")
    sc = SparkContext(conf=conf)

    # make text file into RDD and then map each row by splitting on the tabs
    movie_data = sc.textFile("u.data")
    rows = movie_data.map(lambda line: line.split("\t"))

    # group together all rows by user ID (key) and then make list of all movie they had ranked (value=list)
    idToitem = rows.map(lambda r: (int(r[0]), int(r[1]))).groupByKey()
    print(idToitem.count())
    print(idToitem.map(lambda x: {x[0]: len(x[1])}).collect())

    # lab assignment ask for Id's that have above 25 (at least 26) movie critics
    idRank25 = idToitem.filter(lambda x: len(x[1]) > 25)
    print(idRank25.count())
    print(idRank25.map(lambda x: {x[0]: len(x[1])}).collect())

    # make dictionary of user ID and list of movies, dictionary of user ID and length of list
    # then turn it into RDD to write to file
    id_list = sc.parallelize(idRank25.map(lambda x: {x[0]: list(x[1])}).collect())
    id_len = sc.parallelize(idRank25.map(lambda x: {x[0]: len(x[1])}).collect())

    id_list.saveAsTextFile("IdAndMovieList")
    id_len.saveAsTextFile("IdAndListCount")

    sc.stop()

File: Lab-1-source/CS5542-Lab1-SCALA/src/main/scala/movieRank.scala
import org.apache.spark.{SparkConf, SparkContext}

object movieRank {

  def main(args: Array[String]): Unit = {

    // spark init
    val conf = new SparkConf().setAppName("Movie Ranker").setMaster("local")
    val sc = new SparkContext(conf)

    // load movie data and group movie list by user ID
    val input = sc.textFile("u.data")
    val rows = input.map(line => line.split("\t",4))
      .map(id =>(id(0), id(1)))
      .groupByKey()

    // filter by movie list that is above 25 (at least 26), replace movie list by its length
    val filter25 = rows.filter( _._2.size > 25)
    val filter25_size = filter25.map(id => (id._1, id._2.size))

    // print to console the number of user IDs before and after filter
    System.out.println(rows.count())
    System.out.println(filter25.count())

    // save to file the user id and movie list, user id and movie list length
    filter25.saveAsTextFile("UserId_And_movie_list_output")
    filter25_size.saveAsTextFile("UserId_And_movie_List_length_output")
    System.out.println("Scala Spark Works")
    sc.stop()
  }
}
